### PR TITLE
Support Delta Lake file-based source provider

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: n/a
 - **Parent Issue**: #148 
 - **Dependencies**: n/a

### What changes were proposed in this pull request?

Refactor #197 + #224 (except for hybrid scan test refactoring) based on #227.

This PR introduces DeltaLakeFileBasedSource to support indexes on Delta Lake Sources.

### Does this PR introduce _any_ user-facing change?

Yes,

Users can create/refresh indexes on Delta Lake table and also utilize Hybrid Scan.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by sezruby** [comment]
@imback82 Need to add some comments but could you review this PR?
I'll handle Hybrid Scan test refactoring in #224 with another PR.

**Comment by imback82** [comment]
> @imback82 Need to add some comments but could you review this PR?

Cool! I will take a look. 

> I'll handle Hybrid Scan test refactoring in #224 with another PR.

Yes, I believe Hybrid Scan requires new APIs, so let's handle it separately. Thanks!

**Comment by imback82** [comment]
`DeltaLakeFileBasedSource` looks good to me. If you are OK with overall approach with #227, I can merge it and resolve comments (unit tests) as a follow up PRs.

Btw, can you check the build failure?

**Comment by sezruby** [comment]
> Yes, I believe Hybrid Scan requires new APIs, so let's handle it separately. Thanks!
@imback82 I added `allFiles` API for Hybrid Scan and seems that's all for Hybrid Scan.

BTW #227 also has the same test failure and it's not reproducible on my local. Let me check further.

**Comment by imback82** [comment]
#227 has been merged. Thanks!

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala [hunk 1]
```diff
@@ -70,11 +70,19 @@ private[actions] abstract class RefreshActionBase(
     val latestRelation =
       Hyperspace.getContext(spark).sourceProviderManager.refreshRelation(relations.head)
     val dataSchema = DataType.fromJson(latestRelation.dataSchemaJson).asInstanceOf[StructType]
-    spark.read
-      .schema(dataSchema)
-      .format(latestRelation.fileFormat)
-      .options(latestRelation.options)
-      .load(latestRelation.rootPaths: _*)
+    if (latestRelation.rootPaths.size == 1) {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:73]: Delta Lake only allows one path in load()

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:73]: Yea, we can do something like:
```scala
val df = spark.read
  .schema(dataSchema)
  .format(latestRelation.fileFormat)
  .options(latestRelation.options)
// Due to the difference in how the "path" option is set: https://github.com/apache/spark/blob/ef1441b56c5cab02335d8d2e4ff95cf7e9c9b9ca/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L197,
// load() with a single parameter needs to be handled differently.
if (latestRelation.rootPaths.size == 1) {
  df.load(latestRelation.rootPaths.head)
} else {
  df.load(latestRelation.rootPaths: _*)
}
```

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:73]: Btw, what happens if the delta lake implementation adds the "path" option in `latestRelation.options`?

File: src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala [hunk 2]
```diff
@@ -174,9 +175,14 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
       //    + file:/C:/hyperspace/src/test/part-00003.snappy.parquet
       import spark.implicits._
       val dataPathColumn = "_data_path"
-      val lineageDF = fileIdTracker.getFileToIdMap.toSeq
+      val isDeltaLakeSource = DeltaLakeRuleUtils.isDeltaLakeSource(df.queryExecution.optimizedPlan)
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala:178]: Can we move this to source provider somehow? Basically, we don't want source specific implementation outside the provider.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 3]
```diff
@@ -437,19 +443,24 @@ object RuleUtils {
...
+          // For example, we could have the following in PartitionSpec:
+          //   - partition columns = "col1", "col2"
+          //   - partitions: "/path/col1=1/col2=1", "/path/col1=1/col2=2", etc.
+          // , and going up the same number of directory levels as the number of partition columns
+          // will compute the base path. Note that PartitionSpec.partitions will always contain
+          // all the partitions in the path, so "partitions.head" is taken as an initial value.
+          val basePath = p.partitionSpec.partitionColumns
+            .foldLeft(p.partitionSpec.partitions.head.path)((path, _) => path.getParent)
+          Some(basePath.toString)
+        case d: TahoeLogFileIndex =>
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:461]: same here. we need to push this to provider by introducing new API

File: src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala (hunk 2 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala]: question, why does delta not require the `replace`?

File: src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala:81]: https://github.com/microsoft/hyperspace/pull/265/files#diff-f7ecc68d1799e9c2c916973786081d5f35f312785537ccd2eed61bce41a10786R72
"path" key will be removed in `createRelation`.

File: src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala (hunk 2 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala]: input_file_neme() function of Delta returns "file:/" not "file:///"

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala]: I see. How about we normalize the output of `input_file_name` by removing "file:/" or "file:///"? (and same to the `getFileToIdMap` keys)

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala]: replacing filename - id map is cheaper than normalizing all file paths of all rows. I'll add an API for this.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 4]
```diff
@@ -24,10 +24,12 @@ import org.apache.spark.sql.catalyst.catalog.BucketSpec
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, In, Literal, Not}
 import org.apache.spark.sql.catalyst.optimizer.OptimizeIn
 import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.delta.files.TahoeLogFileIndex
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:27]: --

**Review by imback82 — COMMENTED** [review]
Generally approach/integration looks good! I will do a detailed review this week.

@pirz @apoorvedave1 Could you take a look as well?

**Comment by apoorvedave1** [comment]
LGTM, thanks @sezruby

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            .mode("append")
+            .save(dataPath)
+
+          // The index should be applied for the updated version.
+          assert(verifyIndexUsage(query().queryExecution.optimizedPlan, "deltaIndex", true))
+        }
+      }
+    }
+  }
+
```
Comment (sezruby) [inline src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:147]: Note to reviewers: more hybrid scan + delta lake test cases - https://github.com/microsoft/hyperspace/pull/274

**Review by pirz — COMMENTED** [review]
LGTM, Thanks @sezruby

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  def verifyIndexUsage(
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:148]: nit: `isIndexApplied` or `isIndexUsed`?

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala (hunk 5 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:147]: Is the result comparison tested in #274?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 7]
```diff
@@ -261,7 +262,8 @@ object RuleUtils {
       index: IndexLogEntry,
       plan: LogicalPlan,
       useBucketSpec: Boolean): LogicalPlan = {
-    val isParquetSourceFormat = index.relations.head.fileFormat.equals("parquet")
+    val fileFormat = index.relations.head.fileFormat
+    val isParquetSourceFormat = fileFormat.equals("parquet") || fileFormat.equals("delta")
```
[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:266]: You can do this as a separate PR if that is preferred. Please create an issue in that case.

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala (hunk 5 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:147]: yes `checkAnswer` is used to compare the results in #274

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:266]: I think this case is too specific to create an API to source provider; and it refers fileformat string in index metadata, not relation.
So it's better to create the function in `IndexLogEntry` or some Utils class if needed. WDYT? @imback82

[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:266]: You should be able to plug in a source provider defined externally without changing Hyperspace codebase. For example, let's say I have a source format "blah" that uses parquet internally, and how can I plug in without modifying Hyperspace?



> I think this case is too specific to create an API to source provider; and it refers fileformat string in index metadata, not relation.

You can do this in the create path and record it in the metadata.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeFileBasedSource.scala [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      logicalRelation: LogicalRelation,
+      fileIdTracker: FileIdTracker): Option[Relation] = {
+    logicalRelation.relation match {
+      case HadoopFsRelation(location: TahoeLogFileIndex, _, dataSchema, _, _, options) =>
+        val files = location
+          .getSnapshot(stalenessAcceptable = false)
+          .filesForScan(projection = Nil, location.partitionFilters, keepStats = false)
+          .files
+          .map { f =>
+            new FileStatus(
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeFileBasedSource.scala:64]: Can this also use `toFileStatus`?

**Review by imback82 — COMMENTED** [review]
LGTM except for one pending comment about introducing `hasParquetAsSourceFormat`.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:266]: Got your point - it should be able to add a new source provider externally.
Let me handle this with a new pr & issue. Thanks!

**Review by imback82 — APPROVED** [review]

**Pull request merged**